### ISSUE TITLE
Clarify error message when unpacking objects with `*` and `**`

### ISSUE THREAD

**Issue by cool-RR** [python/cpython#127643]
When attempting to unpack a list (or any other non-mapping) as a mapping, one gets this error message:

```
builtins.TypeError: 'list' object is not a mapping
```

It's confusing, especially since sometimes the unpacking is done on a long multi-line statement, and it's not clear that the unpacking is the source of the error.

I suggest modifying to: 

```
builtins.TypeError: Can't unpack 'list' object using ** because it's not a mapping
```

I'll be happy to submit a pull request if you agree with this change.

**Comment by picnixz** [comment]
I'm not in favor of this change because users do rely on exception messages but I can definitely understand the confusion. Wouldn't it be better if the bad code was highlighted instead of just having  `^^^` carrets? OTOH, can you give an example of where this error occurred?

**Comment by cool-RR** [comment]
> I'm not in favor of this change because users do rely on exception messages but I can definitely understand the confusion. What we could say is "cannot unpack 'list' object using **"

I think it's not as informative as the text I suggested, but maybe other people will chime in. 

>  (you can only unpack mapping-objects so it would be redundant).

Isn't `*` also unpacking?

> Wouldn't it be better if the bad code was highlighted instead of just having `^^^` carrets?

It would be better, but it seems like a separate discussion.

**Comment by picnixz** [comment]
> Isn't * also unpacking?

Yes, but if you mention `**`, then you expect a mapping to be unpacked (that's the redundancy I was talking about).

> I think it's not as informative as the text I suggested, but maybe other people will chime in.

We generally weigh between how much information you need and how much it's needed. in general, we don't specify a reason if that reason can be inferred from the language specifications (in particular, `**` only works on mapping objects)

**Comment by cool-RR** [comment]
> > Isn't * also unpacking?
> 
> Yes, but if you mention `**`, then you expect a mapping to be unpacked (that's the redundancy I was talking about).

I see. However, the text `cannot unpack 'list' object using **` does not convey the core of the problem, which is that `list` does not support the mapping protocol. This is conveyed by the current error message. Even though you and I could easily guess from the text that this is the case, I believe that error messages should also be easily understandble to beginners or people will less experience with Python.

The error message `Cannot unpack 'list' object using ** because it is not a mapping` conveys both pieces of information. If you don't like that it's long, we could shorten it to something like `Cannot unpack non-mapping 'list' object using **`.

**Comment by picnixz** [comment]
Ah, I see your point now. I still think that my message would be sufficient if in addition we have a TypeError, but I agree that I could have been biased. I like your alternative formulation, namely `cannot unpack non-mapping 'list' object using **`.

Asking some docs, typing and core devs on that matter: @JelleZijlstra @AlexWaygood @ncoghlan @willingc @serhiy-storchaka @encukou (or maybe we should open a discourse thread for bikeshedding the error message and/or it would be better if we change that error message).

As a matter of fact, we write `T object is not iterable` when trying to unpack it using `*`, so it could also be helpful to change that one to `cannot *-unpack non-iterable 'T' object` (and `cannot **-unpack non-mapping 'T' object`). Whatever we choose for `**`, we should also reflect it for `*`.

**Comment by cool-RR** [comment]
Great. Agreed regarding the `*` error message.
